metrics/density/docker_memory_usage.sh

```diff
@@ -111,33 +110,48 @@ function get_docker_memory_usage(){
 	# let it do its work and for the numbers to 'settle'.
 	sleep "$WAIT_TIME"
 
-	# Get PSS memory of CC components.
-	# And check that the smem search has found the process - we get a "0"
-	#  back if that procedure fails (such as if a process has changed its name
-	#  or is not running when expected to be so)
-	# As an added bonus - this script must be run as root (or at least as
-	#  a user with enough rights to allow smem to read the smap stats for
-	#  the docker owned processes). Now if you do not have enough rights
-	#  the smem failure to read the stats will also be trapped.
-	qemu_mem="$(get_pss_memory "$QEMU_PATH")"
-	if [ "$qemu_mem" == "0" ]; then
-		die "Failed to find PSS for $QEMU_PATH"
-	fi
+	# Check the runtime in order in order to determine which process will
+	# be measured about PSS
+	if [ "$RUNTIME" == "runc" ]; then
+		runc_workload_mem="$(get_pss_memory "$CMD")"
```

Comment:
Hi.
 Right, I think the notion is potentially right, but a few comments:
- we should add the '0 returned' check like we do below - to check that smem actually found the workload
- I'm slightly nervous that we are checking for such as common command as `sh` (the default CMD we run) - it feels like it would be very easy to have a random `sh` laying around running on the system from something else that would then get included in our results ... but, having said that, we look for `^sh`, and that I think vastly reduces the chances of us picking up a stray `/bin/sh` for instance...
- another idea is maybe we could enter the namespaces of the containers to do the smem to ensure we only pick up the CMDs we are looking for - but, I think we'd have to do that for each container separately, and that feels like a lot of effort for a small gain ;-)
- and finally, we could potentially use `docker stat --no-stream` to get the memory usage numbers directly from docker itself. Having said that, on my local test from smem I get PSS of 392K, and from docker I get a mem usage of 100 KiB - so, who to trust??

Given this is better than what we have, and I suspect in 99% of cases will give us the correct data (not include other random `sh`s) and get that data with the same method we use for CC's (`smem`), I'm probably going to be fine with this - but, @MarioCarrilloA - any thoughts on the above?

Reply:
Hi @grahamwhaley good points, In my opinion we should take the results provided by `smem` (PSS), it seems that `docker stat` shows the information related to the used cgroup, I could not see info related to PSS in documentation
https://docs.docker.com/engine/admin/runmetrics/#memory-metrics-memorystat

And also I agree regarding to looks for `sh`, it will be very easy to "pollute" the results with more random `sh` instances surfing in the system, I will check that :)